src/chainstate/stacks/db/blocks.rs

```diff
@@ -2702,6 +2737,107 @@ impl StacksChainState {
+                .map_err(|e| MemPoolRejection::BadNonces(e))
+        })?;
+
+        // 5: the paying account must have enough funds
+        if fee as u128 > payer.stx_balance {
+            return Err(MemPoolRejection::NotEnoughFunds(fee as u128, payer.stx_balance))
+        }
+
+        // 6: payload-specific checks
+        match &tx.payload {
+            TransactionPayload::TokenTransfer(_addr, amount, _memo) => {
+                // got the funds?
+                let total_spent = (*amount as u128) +
+                    if origin == payer {
+                        fee as u128
+                    } else {
+                        0
+                    };
+                if total_spent > origin.stx_balance {
+                    return Err(MemPoolRejection::NotEnoughFunds(total_spent, origin.stx_balance))
+                }
+            },
+            TransactionPayload::ContractCall(TransactionContractCall {
+                address, contract_name, function_name, function_args }) => {
+                let contract_identifier = QualifiedContractIdentifier::new(address.clone().into(), contract_name.clone());
+
+                let function_type = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+                    conn.with_analysis_db_readonly(|db| {
+                        db.get_public_function_type(&contract_identifier, &function_name)
+                    }) })
+                    .map_err(|_e| MemPoolRejection::NoSuchContract)?
+                    .ok_or_else(|| MemPoolRejection::NoSuchPublicFunction)?;
+
+                let arg_types: Vec<_> = function_args.iter().map(|x| TypeSignature::type_of(x)).collect();
+                function_type.check_args(&arg_types)
+                    .map_err(|e| MemPoolRejection::BadFunctionArgument(e))?;
+            },
+            TransactionPayload::SmartContract(TransactionSmartContract { name, code_body: _ }) => {
+                let contract_identifier = QualifiedContractIdentifier::new(tx.origin_address().into(), name.clone());
+                let exists = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+                    conn.with_analysis_db_readonly(|db| {
+                        db.has_contract(&contract_identifier)
+                    }) });
+                if exists {
+                    return Err(MemPoolRejection::ContractAlreadyExists(contract_identifier))
+                }
+            },
+            TransactionPayload::PoisonMicroblock(microblock_header_1, microblock_header_2) => {
+                if microblock_header_1.sequence != microblock_header_2.sequence ||
+                    microblock_header_1.prev_block != microblock_header_2.prev_block ||
+                    microblock_header_1.version != microblock_header_2.version {
+                    return Err(MemPoolRejection::PoisonMicroblocksDoNotConflict)
+                }
+
+                let anchor_block = &microblock_header_1.prev_block;
```

Comment:
`.prev_block` will be the hash of the previous _microblock_ in the sequence, unless `sequence` is 0.  You'll want to look up the anchor block hash using the `staging_microblocks` table.

Reply:
Hmmm -- that means a node would only ever be able to validate a poison microblock if it had seen the full chain of microblocks for a given block, right?

Reply:
so if there's something like:

```
ub_0 -> ub_1 -> ub_2
    |_> ub_1' -> ub_2'
```

And someone sees `ub_2'` and `ub_2`, then how would a node validate the poison microblock?

Reply:
Almost.  If a node saw _one_ of the equivocated microblocks, and then saw a subsequent poison-microblock transaction, it would be able to validate it (since the signature on the equivocated microblock headers would have come from the anchor block's microblock signing key).

Reply:
Actually, it shouldn't matter if the node saw either of the microblocks at all.  If (1) the node saw the anchor block, and (2) saw that the microblock signing key was used to produce two concurrent microblock headers (which the poison microblock transaction contains), then the node will accept it and punish the leader who produced the anchored block.  Moreover, well-formed poison-microblock transactions _must_ occur within an anchored block (they must have the `OnChainOnly` flag set for their anchor mode).

Reply:
If multiple anchor blocks use the same signing key, then all leaders who produced them are punished.

Reply:
Okay -- so to validate an poison microblock, we'd want to:

1. recover the public key from the signature and microblock header
2. lookup the public key hash in the database to see if its the microblock signing key for any anchor block

If the lookup is successful, the poison microblock transaction is valid.

Reply:
Yes, with the caveat that you'd constrain the query on (2) to only return blocks whose rewards are not yet mature.

Related to that, I'm not sure if we should consider an anchor block invalid if it contains an invalid poison microblock transaction (i.e. one that refers to non-existent microblocks). I'm leaning towards just docking the sender the tx fee and considering the transaction invalid.  Thoughts?

Reply:
I think if a test is easy enough to validate on mempool admission, then any transaction that fails that test shouldn't be included in valid blocks.

Reply:
Addressed in ab40c18